### ISSUE TITLE
Reason for freeze_bn=True in ResNet backbone

### ISSUE THREAD

**Issue by huwenjie333** [fizyr/keras-retinanet#974]
Hello, I'm wondering what's the reason to set `freeze_bn=True` when we create the ResNet Backone as shown in this line: https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/models/resnet.py#L99

Also I see this repo use the `keras-resnet` library instead of ResNet model from official Keras Application library [here](https://keras.io/applications/#resnet). Is it because of this additional `freeze_bn=True` argument, which is not included in Keras Application?

**Comment by DuCheng2018** [comment]
+1. 
I am confused too.
Is this setting caused by using pre-trained model of ImageNet?

**Comment by hgaiser** [comment]
The implementation of resnet in Keras has two flaws: it doesn't follow the original implementation and there is no resnet101 and resnet152 implementations. I don't remember if it was mentioned in a paper or if it is a difference between Caffe in the [original implementation](https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-50-deploy.prototxt) and Keras, but `freeze_bn` is the default in that case.

It also makes sense: allowing the batchnorm layers to update is only valuable if the batch size of the data being passed through is large. Otherwise it adjusts to specific images each iteration, which gives inaccurate updates. For object detection algorithms, the batch size is usually quite low. Note that "group normalization" is designed to work with smaller batches; which is something that would be nice to have at some point in the future.

**Comment by huwenjie333** [comment]
> The implementation of resnet in Keras has two flaws: it doesn't follow the original implementation and there is no resnet101 and resnet152 implementations. I don't remember if it was mentioned in a paper or if it is a difference between Caffe in the [original implementation](https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-50-deploy.prototxt) and Keras, but `freeze_bn` is the default in that case.
> 
> It also makes sense: allowing the batchnorm layers to update is only valuable if the batch size of the data being passed through is large. Otherwise it adjusts to specific images each iteration, which gives inaccurate updates. For object detection algorithms, the batch size is usually quite low. Note that "group normalization" is designed to work with smaller batches; which is something that would be nice to have at some point in the future.

Thanks so much for the reply, when you say "the implementation of resnet", is the one from Keras Application [here](https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet_common.py)? It does provide resnet101 and resnet152, but you are right there is no option to set `freeze_bn`.

**Comment by hgaiser** [comment]
Ahh okay didn't know Keras had implemented the other versions by now,
thanks for sharing! When this was implemented they only had resnet50.

**Comment by quanghainguyen1512** [comment]
> > The implementation of resnet in Keras has two flaws: it doesn't follow the original implementation and there is no resnet101 and resnet152 implementations. I don't remember if it was mentioned in a paper or if it is a difference between Caffe in the [original implementation](https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-50-deploy.prototxt) and Keras, but `freeze_bn` is the default in that case.
> > It also makes sense: allowing the batchnorm layers to update is only valuable if the batch size of the data being passed through is large. Otherwise it adjusts to specific images each iteration, which gives inaccurate updates. For object detection algorithms, the batch size is usually quite low. Note that "group normalization" is designed to work with smaller batches; which is something that would be nice to have at some point in the future.
> 
> Thanks so much for the reply, when you say "the implementation of resnet", is the one from Keras Application [here](https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet_common.py)? It does provide resnet101 and resnet152, but you are right there is no option to set `freeze_bn`.

but currently in keras version 2.2.4 they didn't include them. Those models will be available in next version 2.2.5 as mentioned [here](https://github.com/keras-team/keras/issues/12289)

**Comment by guker** [comment]
> The implementation of resnet in Keras has two flaws: it doesn't follow the original implementation and there is no resnet101 and resnet152 implementations. I don't remember if it was mentioned in a paper or if it is a difference between Caffe in the [original implementation](https://github.com/KaimingHe/deep-residual-networks/blob/master/prototxt/ResNet-50-deploy.prototxt) and Keras, but `freeze_bn` is the default in that case.
> 
> It also makes sense: allowing the batchnorm layers to update is only valuable if the batch size of the data being passed through is large. Otherwise it adjusts to specific images each iteration, which gives inaccurate updates. For object detection algorithms, the batch size is usually quite low. Note that "group normalization" is designed to work with smaller batches; which is something that would be nice to have at some point in the future.

I got it，it‘s great， thanks！

**Comment by wty-yy** [comment]
[This paper](https://arxiv.org/pdf/1906.07155.pdf) (Section 5.2) shows the training result of freezing the related variables in backbone's batch normalization.
